### DISCUSSION TITLE
Why cannot I hook CreateFileW

### DISCUSSION THREAD

**Discussion by 6ziv** [microsoft/Detours#214] [Q&A] [answered]
Well, I am recently on an old game, which requires a CD to play some music.
That's troublesome, and adds a drive label useless to other applications in the global namespace during the game.
By monitoring the application behavior with the Rohitab API Monitor, I found out that the music is played by calling mciSendCommand in mcicda.dll, and mciSendCommand will use CreateFileW to open the drive for reading.
So I decided to hook on the CreateFileW with detours, and redirect the call to a file which is an exact clone of the disk.
I also managed to hook GetDiskFreeSpace(A/W), GetVolumeInformation(A/W),GetLogicalDriveStrings(A/W),GetDriveType(A/W)， and GetLogicalDrives, and they all worked as expected.
However, when hooking on CreateFile(A/W), they successfully redirect most calls by the hooked process, however, the call made by mciSendCommand which I am really interested in does not get into the faked function, though both calls before and after does.
I also tried to delay the hook after LoadLibraryExW, and surprisingly found that the function is calling LoadLibraryExW itself, in a weird way. And again my hook does not catch this LoadLibraryExW, instead, it catches two CreateFileW calls trying to open the dll file when the LoadLibraryExW occurs.
Am I ignoring something, or am I using the library in a wrong way? Or is what I want to do designed to be impossible?
Code is uploaded as attachments. And I can provide the game if needed, but that would be a little large (over 1.5GB). And Google Drive or Mega is extremely slow here (tens or hundreds Kbytes per second).

[HookDll.tar.gz](https://github.com/microsoft/Detours/files/7142083/HookDll.tar.gz)
[RunHookedDll.cpp.gz](https://github.com/microsoft/Detours/files/7142086/RunHookedDll.cpp.gz)

**Comment by Dwedit** [comment, marked as answer]
If you are trying to change a game to not require a CD to play music, there is already a solution for that called "_inmm", which makes the game play MP3 files instead of CD audio tracks.
